pkg/web/oauthclient/callback.go

```diff
@@ -39,8 +38,18 @@ type oauthAuthorizeResponse struct {
 	Code             string `form:"code" query:"code"`
 }
 
-func (res *oauthAuthorizeResponse) ValidateContext(c context.Context) error {
+func (res *oauthAuthorizeResponse) ValidateContext(c echo.Context, oc *OAuthClient) error {
 	if res.Error != "" {
+		if res.Error == "access_denied" {
+			// The `access_denied` error usually means that the authorization request
+			// was actively denied by the user. In that case, it is best to redirect
+			// back to the page root, which will enable to restart the authorization
+			// process if wished.
+			ctx := c.Request().Context()
+			config := oc.configFromContext(ctx)
+			oc.removeStateCookie(c)
+			return c.Redirect(http.StatusFound, config.RootURL)
+		}
```

Comment:
I don't think this code belongs in the validator of the response message. I think it would be better to return a specific access denied error, and matching on that in `HandleCallback`

Reply:
I think there are two occasions where (in our setup) `access_denied` errors happen, which is:

1.  if the authorization was denied in the authorization screen of the Account App
2.  If the OIDC login is canceled

In both situations, it is good to redirect back to the root instead of rendering an error.

Since we don't use error codes in the OAuth server within TTID, we could currently only match the specific error via the error description `User aborted interaction`.